File: Backend/routers/scan.py
import os
import re
from pathlib import Path
from datetime import datetime
import time

from fastapi import APIRouter, HTTPException, UploadFile, File, Form
from fastapi.responses import FileResponse
from bson import ObjectId

from config import settings
from database import get_db

router = APIRouter(prefix="/api/scan", tags=["scan"])

UPLOAD_DIR = Path(settings.upload_dir).resolve()
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

MAX_SIZE = 500 * 1024 * 1024  # 500 MB
ALLOWED_EXT = (".dcm", ".dicom")
ALLOWED_MIME = "application/dicom"

def _allowed_file(filename: str, content_type: str | None) -> bool:
    if filename:
        ext = os.path.splitext(filename)[1].lower()
        if ext in ALLOWED_EXT:
            return True
    if content_type and content_type.strip().lower() == ALLOWED_MIME:
        return True
    return False

@router.get("")
async def list_scans():
    db = get_db()
    cursor = db.scans.find().sort("createdAt", -1)
    scans = []
    async for s in cursor:
        scans.append({
            "id": str(s["_id"]),
            "originalName": s.get("originalName", ""),
            "size": s.get("size", 0),
            "createdAt": s.get("createdAt"),
            "patientName": s.get("patientName", ""),
            "studyDate": s.get("studyDate", ""),
            "modality": s.get("modality", ""),
        })
    return scans

@router.get("/{id}")
async def get_scan(id: str):
    try:
        oid = ObjectId(id)
    except Exception:
        raise HTTPException(404, "Scan not found")
    db = get_db()
    scan = await db.scans.find_one({"_id": oid})
    if not scan:
        raise HTTPException(404, "Scan not found")
    return {
        "id": str(scan["_id"]),
        "originalName": scan.get("originalName", ""),
        "size": scan.get("size", 0),
        "mimeType": scan.get("mimeType", "application/dicom"),
        "createdAt": scan.get("createdAt"),
        "patientName": scan.get("patientName", ""),
        "studyDate": scan.get("studyDate", ""),
        "modality": scan.get("modality", ""),
    }

@router.get("/{id}/file")
async def get_scan_file(id: str):
    try:
        oid = ObjectId(id)
    except Exception:
        raise HTTPException(404, "Scan not found")
    db = get_db()
    scan = await db.scans.find_one({"_id": oid})
    if not scan:
        raise HTTPException(404, "Scan not found")
    path = Path(scan["path"])
    if not path.is_absolute():
        path = UPLOAD_DIR / path
    else:
        path = Path(scan["path"])
    if not path.exists():
        raise HTTPException(404, "File not found on disk")
    try:
        path = path.resolve()
        upload_resolved = UPLOAD_DIR.resolve()
        path.relative_to(upload_resolved)
    except ValueError:
        raise HTTPException(403, "Invalid path")
    return FileResponse(
        path,
        media_type=scan.get("mimeType") or "application/dicom",
    )

@router.delete("/{id}")
async def delete_scan(id: str):
    try:
        oid = ObjectId(id)
    except Exception:
        raise HTTPException(404, "Scan not found")
    db = get_db()
    scan = await db.scans.find_one({"_id": oid})
    if not scan:
        raise HTTPException(404, "Scan not found")
    path = Path(scan["path"])
    if not path.is_absolute():
        path = UPLOAD_DIR / path
    else:
        path = Path(scan["path"])
    if path.exists():
        try:
            path.unlink()
        except Exception:
            pass
    await db.scans.delete_one({"_id": oid})
    return {"ok": True}

@router.post("/upload")
async def upload_scan(
    file: UploadFile = File(...),
    patientName: str = Form(""),
    studyDate: str = Form(""),
    modality: str = Form(""),
):
    if not file.filename and not file.content_type:
        raise HTTPException(400, 'No DICOM file uploaded. Use form field "file".')
    if not _allowed_file(file.filename or "", file.content_type):
        raise HTTPException(400, "Only DICOM files (.dcm, .dicom) are allowed")
    content = await file.read()
    if len(content) > MAX_SIZE:
        raise HTTPException(400, "File too large")

    pname = (patientName or "").strip()
    sdate = (studyDate or "").strip()
    mod = (modality or "").strip()
    orig = file.filename or "scan"

    db = get_db()
    existing = await db.scans.find_one({"originalName": orig, "studyDate": sdate})
    if existing and (existing.get("patientName") or "").lower() == (pname or "").lower():
        from fastapi.responses import JSONResponse
        return JSONResponse(
            status_code=201,
            content={
                "ok": True,
                "uploaded": [],
                "skipped": True,
                "reason": "A scan with the same patient, study date, and file name already exists.",
            },
        )

    safe_name = re.sub(r"[^a-zA-Z0-9._-]", "_", orig)
    safe_name = f"{int(time.time() * 1000)}-{safe_name}"
    dest = UPLOAD_DIR / safe_name
    dest.write_bytes(content)
    doc = {
        "originalName": orig,
        "path": str(dest.resolve()),
        "size": len(content),
        "mimeType": file.content_type or "application/dicom",
        "createdAt": datetime.utcnow(),
        "patientName": pname,
        "studyDate": sdate,
        "modality": mod,
    }
    r = await db.scans.insert_one(doc)
    return {
        "ok": True,
        "uploaded": [
            {
                "id": str(r.inserted_id),
                "originalName": doc["originalName"],
                "size": doc["size"],
                "createdAt": doc["createdAt"],
                "patientName": doc["patientName"],
                "studyDate": doc["studyDate"],
                "modality": doc["modality"],
            }
        ],
    }
}

File: Backend/routes/scan.js
import express from 'express';
import multer from 'multer';
import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';
import Scan from '../models/Scan.js';
import { authMiddleware } from '../middleware/auth.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const UPLOAD_DIR = path.join(__dirname, '../uploads/dicom');

// Ensure upload directory exists
if (!fs.existsSync(UPLOAD_DIR)) {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

// Configure multer
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, UPLOAD_DIR);
  },
  filename: (req, file, cb) => {
    const timestamp = Date.now();
    const originalName = file.originalname.replace(/[^a-zA-Z0-9._-]/g, '_');
    cb(null, `${timestamp}-${originalName}`);
  }
});

const upload = multer({
  storage,
  limits: { fileSize: 500 * 1024 * 1024 }, // 500 MB
  fileFilter: (req, file, cb) => {
    const allowedExtensions = ['.dcm', '.dicom'];
    const ext = path.extname(file.originalname).toLowerCase();
    
    if (allowedExtensions.includes(ext) || file.mimetype === 'application/dicom') {
      cb(null, true);
    } else {
      cb(new Error('Only DICOM files (.dcm, .dicom) are allowed'));
    }
  }
});

const router = express.Router();

// Apply auth middleware to all routes
router.use(authMiddleware);

// Get all scans for authenticated user
router.get('/', async (req, res) => {
  try {
    const scans = await Scan.find({ userId: req.user._id })
      .sort({ createdAt: -1 })
      .lean();

    res.json({
      success: true,
      count: scans.length,
      scans: scans.map(scan => ({
        id: scan._id,
        originalName: scan.originalName,
        size: scan.size,
        createdAt: scan.createdAt,
        patientName: scan.patientName || '',
        studyDate: scan.studyDate || '',
        modality: scan.modality || '',
        doctorName: scan.doctorName || ''
      }))
    });
  } catch (err) {
    console.error('❌ Get scans error:', err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Failed to list scans'
    });
  }
});

// Get specific scan metadata
router.get('/:id', async (req, res) => {
  try {
    const scan = await Scan.findOne({
      _id: req.params.id,
      userId: req.user._id
    }).lean();

    if (!scan) {
      return res.status(404).json({
        success: false,
        error: 'Scan not found'
      });
    }

    res.json({
      success: true,
      scan: {
        id: scan._id,
        originalName: scan.originalName,
        size: scan.size,
        mimeType: scan.mimeType,
        createdAt: scan.createdAt,
        patientName: scan.patientName || '',
        studyDate: scan.studyDate || '',
        modality: scan.modality || '',
        doctorName: scan.doctorName || ''
      }
    });
  } catch (err) {
    console.error('❌ Get scan error:', err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Failed to get scan'
    });
  }
});

// Download DICOM file
router.get('/:id/file', async (req, res) => {
  try {
    const scan = await Scan.findOne({
      _id: req.params.id,
      userId: req.user._id
    });

    if (!scan) {
      return res.status(404).json({
        success: false,
        error: 'Scan not found'
      });
    }

    const filePath = path.resolve(scan.path);
    const uploadDirResolved = path.resolve(UPLOAD_DIR);

    // Security: ensure file is within upload directory
    if (!filePath.startsWith(uploadDirResolved)) {
      return res.status(403).json({
        success: false,
        error: 'Access denied'
      });
    }

    // Check if file exists
    if (!fs.existsSync(filePath)) {
      return res.status(404).json({
        success: false,
        error: 'File not found on disk'
      });
    }

    res.setHeader('Content-Type', scan.mimeType || 'application/dicom');
    res.setHeader('Content-Disposition', `attachment; filename="${scan.originalName}"`);
    res.sendFile(filePath);
  } catch (err) {
    console.error('❌ Download file error:', err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Failed to download file'
    });
  }
});

// Upload DICOM file
router.post('/upload', upload.single('file'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({
        success: false,
        error: 'No DICOM file uploaded. Use form field "file"'
      });
    }

    const patientName = req.body?.patientName ? String(req.body.patientName).trim() : '';
    const studyDate = req.body?.studyDate ? String(req.body.studyDate).trim() : '';
    const modality = req.body?.modality ? String(req.body.modality).trim() : '';

    // Check for duplicates
    const existingScan = await Scan.findOne({
      userId: req.user._id,
      originalName: req.file.originalname,
      studyDate: studyDate || '',
      patientName: { $regex: `^${patientName}$`, $options: 'i' }
    }).lean();

    if (existingScan) {
      // Delete uploaded file since it's a duplicate
      fs.unlinkSync(req.file.path);
      
      return res.status(201).json({
        success: true,
        message: 'Scan already exists',
        skipped: true,
        reason: 'A scan with the same patient, study date, and file name already exists'
      });
    }

    // Create scan record
    const scan = await Scan.create({
      userId: req.user._id,
      originalName: req.file.originalname,
      path: req.file.path,
      size: req.file.size,
      mimeType: req.file.mimetype || 'application/dicom',
      doctorName: req.user.name,
      patientName,
      studyDate,
      modality
    });

    console.log(`✅ DICOM file uploaded: ${scan.originalName}`);

    res.status(201).json({
      success: true,
      message: 'DICOM file uploaded successfully',
      scan: {
        id: scan._id,
        originalName: scan.originalName,
        size: scan.size,
        createdAt: scan.createdAt,
        patientName: scan.patientName,
        studyDate: scan.studyDate,
        modality: scan.modality
      }
    });
  } catch (err) {
    // Clean up uploaded file on error
    if (req.file && fs.existsSync(req.file.path)) {
      fs.unlinkSync(req.file.path);
    }

    console.error('❌ Upload error:', err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Failed to upload file'
    });
  }
});

// Delete scan
router.delete('/:id', async (req, res) => {
  try {
    const scan = await Scan.findOne({
      _id: req.params.id,
      userId: req.user._id
    });

    if (!scan) {
      return res.status(404).json({
        success: false,
        error: 'Scan not found'
      });
    }

    const filePath = path.resolve(scan.path);
    const uploadDirResolved = path.resolve(UPLOAD_DIR);

    // Security check
    if (filePath.startsWith(uploadDirResolved) && fs.existsSync(filePath)) {
      try {
        fs.unlinkSync(filePath);
      } catch (err) {
        console.warn('⚠️  Could not delete file:', err.message);
      }
    }

    await Scan.deleteOne({ _id: req.params.id, userId: req.user._id });

    console.log(`✅ Scan deleted: ${scan.originalName}`);

    res.json({
      success: true,
      message: 'Scan deleted successfully'
    });
  } catch (err) {
    console.error('❌ Delete error:', err.message);
    res.status(500).json({
      success: false,
      error: err.message || 'Failed to delete scan'
    });
  }
});

export default router;
